src/Workspaces/Core/Portable/Diagnostics/HostDiagnosticAnalyzers.cs

```diff
@@ -46,30 +47,31 @@ internal sealed class HostDiagnosticAnalyzers
         /// </summary>
         private ImmutableDictionary<string, DiagnosticAnalyzer> _compilerDiagnosticAnalyzerMap;
 
-        public HostDiagnosticAnalyzers(Lazy<ImmutableArray<HostDiagnosticAnalyzerPackage>> hostAnalyzerPackages, IAnalyzerAssemblyLoader? hostAnalyzerAssemblyLoader, AbstractHostDiagnosticUpdateSource? hostDiagnosticUpdateSource, PrimaryWorkspace primaryWorkspace)
-            : this(new Lazy<ImmutableArray<AnalyzerReference>>(() => CreateAnalyzerReferencesFromPackages(hostAnalyzerPackages.Value, new HostAnalyzerReferenceDiagnosticReporter(hostDiagnosticUpdateSource, primaryWorkspace), hostAnalyzerAssemblyLoader), isThreadSafe: true))
-        {
-        }
-
-        internal HostDiagnosticAnalyzers(ImmutableArray<AnalyzerReference> hostAnalyzerReferences)
-            : this(new Lazy<ImmutableArray<AnalyzerReference>>(() => hostAnalyzerReferences))
-        {
-        }
-
-        private HostDiagnosticAnalyzers(Lazy<ImmutableArray<AnalyzerReference>> hostAnalyzerReferences)
+        /// <summary>
+        /// Maps list of analyzer references and <see cref="LanguageNames"/> to <see cref="SkippedHostAnalyzersInfo"/>.
+        /// </summary>
+        /// <remarks>
+        /// TODO: https://github.com/dotnet/roslyn/issues/42848
+        /// It is quite common for multiple projects to have the same set of analyzer references, yet we will create
+        /// multiple instances of the analyzer list and thus not share the info.
+        /// </remarks>
+        private ConditionalWeakTable<IReadOnlyList<AnalyzerReference>, StrongBox<ImmutableDictionary<string, SkippedHostAnalyzersInfo>>> _skippedHostAnalyzers;
```

Comment:
Is there any risk of different instances being passed in to this that have the same values? Say the underlying thing is an ImmutableArray and it gets boxed multiple times in different instances? We've done other hacks like this but it always felt risky...

Reply:
Yes, that could happen but it doesn't matter for correctness. This is just a cache. We do reuse boxed immutable arrays through out the solution APIs, so we should be good. In any case, this is not ideal and we can do better. See the comment above and https://github.com/dotnet/roslyn/issues/42848